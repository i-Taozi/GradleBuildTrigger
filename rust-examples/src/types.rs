///类型
pub fn types() {
    let decimal = 65.4321_f32;
    //不支持隐式转化
    //let integer: u8 = decimal;
    // FIXME ^ Comment out this line

    //显示转化
    let integer = decimal as u8;
    let character = integer as char;

    println!("Casting: {} -> {} -> {}", decimal, integer, character);

    // 1000 already fits in a u16
    println!("1000 as a u16 is: {}", 1000 as u16);

    // 1000 - 256 - 256 - 256 = 232
    // Under the hood, the first 8 least significant bits (LSB) are kept,
    // while the rest towards the most significant bit (MSB) get truncated.
    //println!("1000 as a u8 is : {}", 1000 as u8);
    // -1 + 256 = 255
    println!("  -1 as a u8 is : {}", (-1i8) as u8);
    println!("1000 mod 256 is : {}", 1000 % 256);

    println!(" 128 as a i16 is: {}", 128 as i16);
    //println!(" 128 as a i8 is : {}", 128 as i8);
    //println!("1000 as a u8 is : {}", 1000 as u8);
    //println!(" 232 as a i8 is : {}", 232 as i8);

    literals();
}

fn literals() {
    //类型作为后缀
    let x = 1u8;
    let y = 2u32;
    let z = 3f32;

    //非固定文字，其类型取决于如何使用
    let i = 1;
    let f = 1.0;

    //Returns the size of the pointed-to value in bytes.
    println!("size of `x` in bytes: {}", std::mem::size_of_val(&x));
    println!("size of `y` in bytes: {}", std::mem::size_of_val(&y));
    println!("size of `z` in bytes: {}", std::mem::size_of_val(&z));
    println!("size of `i` in bytes: {}", std::mem::size_of_val(&i));
    println!("size of `f` in bytes: {}", std::mem::size_of_val(&f));
}
