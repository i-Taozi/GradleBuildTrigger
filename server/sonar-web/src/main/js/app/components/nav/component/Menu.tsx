/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import classNames from 'classnames';
import { LocationDescriptorObject } from 'history';
import { omit } from 'lodash';
import * as React from 'react';
import { Link, LinkProps } from 'react-router';
import Dropdown from '../../../../components/controls/Dropdown';
import Tooltip from '../../../../components/controls/Tooltip';
import { withAppState } from '../../../../components/hoc/withAppState';
import BulletListIcon from '../../../../components/icons/BulletListIcon';
import DropdownIcon from '../../../../components/icons/DropdownIcon';
import NavBarTabs from '../../../../components/ui/NavBarTabs';
import { getBranchLikeQuery, isPullRequest } from '../../../../helpers/branch-like';
import { hasMessage, translate, translateWithParameters } from '../../../../helpers/l10n';
import { getPortfolioUrl, getProjectQueryUrl } from '../../../../helpers/urls';
import { BranchLike, BranchParameters } from '../../../../types/branch-like';
import { ComponentQualifier, isPortfolioLike } from '../../../../types/component';
import './Menu.css';

const SETTINGS_URLS = [
  '/application/console',
  '/project/admin',
  '/project/baseline',
  '/project/branches',
  '/project/settings',
  '/project/quality_profiles',
  '/project/quality_gate',
  '/project/links',
  '/project_roles',
  '/project/history',
  'background_tasks',
  '/project/key',
  '/project/deletion',
  '/project/webhooks'
];

interface Props {
  appState: Pick<T.AppState, 'branchesEnabled'>;
  branchLike: BranchLike | undefined;
  branchLikes: BranchLike[] | undefined;
  component: T.Component;
  isInProgress?: boolean;
  isPending?: boolean;
  onToggleProjectInfo: () => void;
}

type Query = BranchParameters & { id: string };

export class Menu extends React.PureComponent<Props> {
  hasAnalysis = () => {
    const { branchLikes = [], component, isInProgress, isPending } = this.props;
    const hasBranches = branchLikes.length > 1;
    return hasBranches || isInProgress || isPending || component.analysisDate !== undefined;
  };

  isProject = () => {
    return this.props.component.qualifier === ComponentQualifier.Project;
  };

  isDeveloper = () => {
    return this.props.component.qualifier === ComponentQualifier.Developper;
  };

  isPortfolio = () => {
    const { qualifier } = this.props.component;
    return isPortfolioLike(qualifier);
  };

  isApplication = () => {
    return this.props.component.qualifier === ComponentQualifier.Application;
  };

  isAllChildProjectAccessible = () => {
    return Boolean(this.props.component.canBrowseAllChildProjects);
  };

  isApplicationChildInaccessble = () => {
    return this.isApplication() && !this.isAllChildProjectAccessible();
  };

  isGovernanceEnabled = () => {
    const {
      component: { extensions }
    } = this.props;

    return extensions && extensions.some(extension => extension.key.startsWith('governance/'));
  };

  getConfiguration = () => {
    return this.props.component.configuration || {};
  };

  getQuery = (): Query => {
    return { id: this.props.component.key, ...getBranchLikeQuery(this.props.branchLike) };
  };

  renderLinkWhenInaccessibleChild(label: React.ReactNode) {
    return (
      <li>
        <Tooltip
          overlay={translateWithParameters(
            'layout.all_project_must_be_accessible',
            translate('qualifier', this.props.component.qualifier)
          )}>
          <a aria-disabled="true" className="disabled-link">
            {label}
          </a>
        </Tooltip>
      </li>
    );
  }

  renderMenuLink = ({
    label,
    to,
    ...props
  }: Omit<LinkProps, 'to'> & {
    label: React.ReactNode;
    to: LocationDescriptorObject;
  }) => {
    const hasAnalysis = this.hasAnalysis();
    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();
    const query = this.getQuery();
    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(label);
    }
    return (
      <li>
        {hasAnalysis ? (
          <Link to={{ ...to, query: { ...query, ...to.query } }} {...omit(props, ['to'])}>
            {label}
          </Link>
        ) : (
          <Tooltip overlay={translate('layout.must_be_configured')}>
            <a aria-disabled="true" className="disabled-link">
              {label}
            </a>
          </Tooltip>
        )}
      </li>
    );
  };

  renderDashboardLink = () => {
    const { id, ...branchLike } = this.getQuery();

    if (this.isPortfolio()) {
      return this.isGovernanceEnabled() ? (
        <li>
          <Link activeClassName="active" to={getPortfolioUrl(id)}>
            {translate('overview.page')}
          </Link>
        </li>
      ) : null;
    }

    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();
    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(translate('overview.page'));
    }
    return (
      <li>
        <Link activeClassName="active" to={getProjectQueryUrl(id, branchLike)}>
          {translate('overview.page')}
        </Link>
      </li>
    );
  };

  renderCodeLink = () => {
    const isPortfolio = this.isPortfolio();
    const isApplication = this.isApplication();
    const label =
      isPortfolio || isApplication ? translate('view_projects.page') : translate('code.page');
    if (this.isDeveloper()) {
      return null;
    }

    return this.renderMenuLink({ label, to: { pathname: '/code' } });
  };

  renderActivityLink = () => {
    const { branchLike } = this.props;

    if (isPullRequest(branchLike)) {
      return null;
    }

    return this.renderMenuLink({
      label: translate('project_activity.page'),
      to: { pathname: '/project/activity' }
    });
  };

  renderIssuesLink = () => {
    return this.renderMenuLink({
      label: translate('issues.page'),
      to: { pathname: '/project/issues', query: { resolved: 'false' } }
    });
  };

  renderComponentMeasuresLink = () => {
    return this.renderMenuLink({
      label: translate('layout.measures'),
      to: { pathname: 'component_measures' }
    });
  };

  renderSecurityHotspotsLink = () => {
    const isPortfolio = this.isPortfolio();
    return (
      !isPortfolio &&
      this.renderMenuLink({
        label: translate('layout.security_hotspots'),
        to: { pathname: '/security_hotspots' }
      })
    );
  };

  renderSecurityReports = () => {
    const { branchLike, component } = this.props;
    const { extensions = [] } = component;

    if (isPullRequest(branchLike)) {
      return null;
    }

    const hasSecurityReportsEnabled = extensions.some(extension =>
      extension.key.startsWith('securityreport/')
    );

    if (!hasSecurityReportsEnabled) {
      return null;
    }

    return this.renderMenuLink({
      label: translate('layout.security_reports'),
      to: { pathname: '/project/extension/securityreport/securityreport' }
    });
  };

  renderAdministration = () => {
    const { branchLike, component } = this.props;
    const isProject = this.isProject();
    const isPortfolio = this.isPortfolio();
    const isApplication = this.isApplication();
    const query = this.getQuery();

    if (!this.getConfiguration().showSettings || isPullRequest(branchLike)) {
      return null;
    }

    const isSettingsActive = SETTINGS_URLS.some(url => window.location.href.indexOf(url) !== -1);

    const adminLinks = this.renderAdministrationLinks(query, isProject, isApplication, isPortfolio);
    if (!adminLinks.some(link => link != null)) {
      return null;
    }

    return (
      <Dropdown
        data-test="administration"
        overlay={<ul className="menu">{adminLinks}</ul>}
        tagName="li">
        {({ onToggleClick, open }) => (
          <a
            aria-expanded={open}
            aria-haspopup="menu"
            role="button"
            className={classNames('dropdown-toggle', { active: isSettingsActive || open })}
            href="#"
            id="component-navigation-admin"
            onClick={onToggleClick}>
            {hasMessage('layout.settings', component.qualifier)
              ? translate('layout.settings', component.qualifier)
              : translate('layout.settings')}
            <DropdownIcon className="little-spacer-left" />
          </a>
        )}
      </Dropdown>
    );
  };

  renderAdministrationLinks = (
    query: Query,
    isProject: boolean,
    isApplication: boolean,
    isPortfolio: boolean
  ) => {
    return [
      this.renderSettingsLink(query, isApplication, isPortfolio),
      this.renderBranchesLink(query, isProject),
      this.renderBaselineLink(query, isApplication, isPortfolio),
      this.renderConsoleAppLink(query, isApplication),
      this.renderReportSettingsLink(query, isApplication),
      ...this.renderAdminExtensions(query, isApplication),
      this.renderImportExportLink(query, isProject),
      this.renderProfilesLink(query),
      this.renderQualityGateLink(query),
      this.renderLinksLink(query),
      this.renderPermissionsLink(query),
      this.renderBackgroundTasksLink(query),
      this.renderUpdateKeyLink(query),
      this.renderWebhooksLink(query, isProject),
      this.renderDeletionLink(query)
    ];
  };

  renderProjectInformationButton = () => {
    const isProject = this.isProject();
    const isApplication = this.isApplication();
    const label = translate(isProject ? 'project' : 'application', 'info.title');
    const isApplicationChildInaccessble = this.isApplicationChildInaccessble();

    if (isPullRequest(this.props.branchLike)) {
      return null;
    }

    if (isApplicationChildInaccessble) {
      return this.renderLinkWhenInaccessibleChild(label);
    }

    return (
      (isProject || isApplication) && (
        <li>
          <a
            className="menu-button"
            onClick={(e: React.SyntheticEvent<HTMLAnchorElement>) => {
              e.preventDefault();
              e.currentTarget.blur();
              this.props.onToggleProjectInfo();
            }}
            role="button"
            tabIndex={0}>
            <BulletListIcon className="little-spacer-right" />
            {label}
          </a>
        </li>
      )
    );
  };

  renderSettingsLink = (query: Query, isApplication: boolean, isPortfolio: boolean) => {
    if (!this.getConfiguration().showSettings || isApplication || isPortfolio) {
      return null;
    }
    return (
      <li key="settings">
        <Link activeClassName="active" to={{ pathname: '/project/settings', query }}>
          {translate('project_settings.page')}
        </Link>
      </li>
    );
  };

  renderBranchesLink = (query: Query, isProject: boolean) => {
    if (
      !this.props.appState.branchesEnabled ||
      !isProject ||
      !this.getConfiguration().showSettings
    ) {
      return null;
    }

    return (
      <li key="branches">
        <Link activeClassName="active" to={{ pathname: '/project/branches', query }}>
          {translate('project_branch_pull_request.page')}
        </Link>
      </li>
    );
  };

  renderBaselineLink = (query: Query, isApplication: boolean, isPortfolio: boolean) => {
    if (!this.getConfiguration().showSettings || isApplication || isPortfolio) {
      return null;
    }
    return (
      <li key="baseline">
        <Link activeClassName="active" to={{ pathname: '/project/baseline', query }}>
          {translate('project_baseline.page')}
        </Link>
      </li>
    );
  };

  renderConsoleAppLink = (query: Query, isApplication: boolean) => {
    if (!isApplication) {
      return null;
    }
    return (
      <li key="app-console">
        <Link activeClassName="active" to={{ pathname: '/application/console', query }}>
          {translate('application_console.page')}
        </Link>
      </li>
    );
  };

  renderReportSettingsLink = (query: Query, isApplication: boolean) => {
    const extensions = this.getConfiguration().extensions || [];
    const hasGovernance = extensions.find(e => e.key === 'governance/console');

    if (!isApplication || !hasGovernance) {
      return null;
    }

    return (
      <li key="report-settings">
        <Link activeClassName="active" to={{ pathname: '/application/settings', query }}>
          {translate('application_settings.report')}
        </Link>
      </li>
    );
  };

  renderImportExportLink = (query: Query, isProject: boolean) => {
    if (!isProject) {
      return null;
    }
    return (
      <li key="import-export">
        <Link activeClassName="active" to={{ pathname: '/project/import_export', query }}>
          {translate('project_dump.page')}
        </Link>
      </li>
    );
  };

  renderProfilesLink = (query: Query) => {
    if (!this.getConfiguration().showQualityProfiles) {
      return null;
    }
    return (
      <li key="profiles">
        <Link activeClassName="active" to={{ pathname: '/project/quality_profiles', query }}>
          {translate('project_quality_profiles.page')}
        </Link>
      </li>
    );
  };

  renderQualityGateLink = (query: Query) => {
    if (!this.getConfiguration().showQualityGates) {
      return null;
    }
    return (
      <li key="quality_gate">
        <Link activeClassName="active" to={{ pathname: '/project/quality_gate', query }}>
          {translate('project_quality_gate.page')}
        </Link>
      </li>
    );
  };

  renderLinksLink = (query: Query) => {
    if (!this.getConfiguration().showLinks) {
      return null;
    }
    return (
      <li key="links">
        <Link activeClassName="active" to={{ pathname: '/project/links', query }}>
          {translate('project_links.page')}
        </Link>
      </li>
    );
  };

  renderPermissionsLink = (query: Query) => {
    if (!this.getConfiguration().showPermissions) {
      return null;
    }
    return (
      <li key="permissions">
        <Link activeClassName="active" to={{ pathname: '/project_roles', query }}>
          {translate('permissions.page')}
        </Link>
      </li>
    );
  };

  renderBackgroundTasksLink = (query: Query) => {
    if (!this.getConfiguration().showBackgroundTasks) {
      return null;
    }
    return (
      <li key="background_tasks">
        <Link activeClassName="active" to={{ pathname: '/project/background_tasks', query }}>
          {translate('background_tasks.page')}
        </Link>
      </li>
    );
  };

  renderUpdateKeyLink = (query: Query) => {
    if (!this.getConfiguration().showUpdateKey) {
      return null;
    }
    return (
      <li key="update_key">
        <Link activeClassName="active" to={{ pathname: '/project/key', query }}>
          {translate('update_key.page')}
        </Link>
      </li>
    );
  };

  renderWebhooksLink = (query: Query, isProject: boolean) => {
    if (!this.getConfiguration().showSettings || !isProject) {
      return null;
    }
    return (
      <li key="webhooks">
        <Link activeClassName="active" to={{ pathname: '/project/webhooks', query }}>
          {translate('webhooks.page')}
        </Link>
      </li>
    );
  };

  renderDeletionLink = (query: Query) => {
    const { qualifier } = this.props.component;

    if (!this.getConfiguration().showSettings) {
      return null;
    }

    if (
      ![
        ComponentQualifier.Project,
        ComponentQualifier.Portfolio,
        ComponentQualifier.Application
      ].includes(qualifier as ComponentQualifier)
    ) {
      return null;
    }

    return (
      <li key="project_delete">
        <Link activeClassName="active" to={{ pathname: '/project/deletion', query }}>
          {translate('deletion.page')}
        </Link>
      </li>
    );
  };

  renderExtension = ({ key, name }: T.Extension, isAdmin: boolean, baseQuery: Query) => {
    const pathname = isAdmin ? `/project/admin/extension/${key}` : `/project/extension/${key}`;
    const query = { ...baseQuery, qualifier: this.props.component.qualifier };
    return (
      <li key={key}>
        <Link activeClassName="active" to={{ pathname, query }}>
          {name}
        </Link>
      </li>
    );
  };

  renderAdminExtensions = (query: Query, isApplication: boolean) => {
    const extensions = this.getConfiguration().extensions || [];
    return extensions
      .filter(e => !isApplication || e.key !== 'governance/console')
      .map(e => this.renderExtension(e, true, query));
  };

  renderExtensions = () => {
    const query = this.getQuery();
    const extensions = this.props.component.extensions || [];
    const withoutSecurityExtension = extensions.filter(
      extension =>
        !extension.key.startsWith('securityreport/') && !extension.key.startsWith('governance/')
    );

    if (withoutSecurityExtension.length === 0) {
      return null;
    }

    return (
      <Dropdown
        data-test="extensions"
        overlay={
          <ul className="menu">
            {withoutSecurityExtension.map(e => this.renderExtension(e, false, query))}
          </ul>
        }
        tagName="li">
        {({ onToggleClick, open }) => (
          <a
            aria-expanded={open}
            aria-haspopup="menu"
            role="button"
            className={classNames('dropdown-toggle', { active: open })}
            href="#"
            id="component-navigation-more"
            onClick={onToggleClick}>
            {translate('more')}
            <DropdownIcon className="little-spacer-left" />
          </a>
        )}
      </Dropdown>
    );
  };

  render() {
    return (
      <div className="display-flex-center display-flex-space-between">
        <NavBarTabs>
          {this.renderDashboardLink()}
          {this.renderIssuesLink()}
          {this.renderSecurityHotspotsLink()}
          {this.renderSecurityReports()}
          {this.renderComponentMeasuresLink()}
          {this.renderCodeLink()}
          {this.renderActivityLink()}
          {this.renderExtensions()}
        </NavBarTabs>
        <NavBarTabs>
          {this.renderAdministration()}
          {this.renderProjectInformationButton()}
        </NavBarTabs>
      </div>
    );
  }
}

export default withAppState(Menu);
