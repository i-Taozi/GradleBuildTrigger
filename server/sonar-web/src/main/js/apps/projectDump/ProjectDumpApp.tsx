/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { getActivity } from '../../api/ce';
import { getStatus } from '../../api/project-dump';
import throwGlobalError from '../../app/utils/throwGlobalError';
import { withAppState } from '../../components/hoc/withAppState';
import { translate } from '../../helpers/l10n';
import { DumpStatus, DumpTask } from '../../types/project-dump';
import { TaskStatuses, TaskTypes } from '../../types/tasks';
import Export from './components/Export';
import Import from './components/Import';
import './styles.css';

const POLL_INTERNAL = 5000;

interface Props {
  appState: Pick<T.AppState, 'projectImportFeatureEnabled'>;
  component: T.Component;
}

interface State {
  lastAnalysisTask?: DumpTask;
  lastExportTask?: DumpTask;
  lastImportTask?: DumpTask;
  status?: DumpStatus;
}

export class ProjectDumpApp extends React.Component<Props, State> {
  mounted = false;
  state: State = {};

  componentDidMount() {
    this.mounted = true;
    this.loadStatus();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.component.key !== this.props.component.key) {
      this.loadStatus();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  getLastTask(component: string, type: TaskTypes) {
    const data = {
      type,
      component,
      onlyCurrents: true,
      status: [
        TaskStatuses.Pending,
        TaskStatuses.InProgress,
        TaskStatuses.Success,
        TaskStatuses.Failed,
        TaskStatuses.Canceled
      ].join(',')
    };
    return getActivity(data)
      .then(({ tasks }) => (tasks.length > 0 ? tasks[0] : undefined), throwGlobalError)
      .catch(() => undefined);
  }

  getLastTaskOfEachType(componentKey: string) {
    const {
      appState: { projectImportFeatureEnabled }
    } = this.props;

    const all = projectImportFeatureEnabled
      ? [
          this.getLastTask(componentKey, TaskTypes.ProjectExport),
          this.getLastTask(componentKey, TaskTypes.ProjectImport),
          this.getLastTask(componentKey, TaskTypes.Report)
        ]
      : [
          this.getLastTask(componentKey, TaskTypes.ProjectExport),
          Promise.resolve(),
          this.getLastTask(componentKey, TaskTypes.Report)
        ];
    return Promise.all(all).then(([lastExportTask, lastImportTask, lastAnalysisTask]) => ({
      lastExportTask,
      lastImportTask,
      lastAnalysisTask
    }));
  }

  loadStatus = () => {
    const { component } = this.props;
    return Promise.all([getStatus(component.key), this.getLastTaskOfEachType(component.key)]).then(
      ([status, { lastExportTask, lastImportTask, lastAnalysisTask }]) => {
        if (this.mounted) {
          this.setState({
            status,
            lastExportTask,
            lastImportTask,
            lastAnalysisTask
          });
        }
        return {
          status,
          lastExportTask,
          lastImportTask,
          lastAnalysisTask
        };
      }
    );
  };

  poll = () => {
    this.loadStatus().then(
      ({ lastExportTask, lastImportTask }) => {
        if (this.mounted) {
          const progressStatus = [TaskStatuses.Pending, TaskStatuses.InProgress];
          const exportNotFinished =
            lastExportTask === undefined || progressStatus.includes(lastExportTask.status);
          const importNotFinished =
            lastImportTask === undefined || progressStatus.includes(lastImportTask.status);
          if (exportNotFinished || importNotFinished) {
            setTimeout(this.poll, POLL_INTERNAL);
          } else {
            // Since we fetch status separate from task we could not get an up to date status.
            // even if we detect that export / import is finish.
            // Doing a last call will make sur we get the latest status.
            this.loadStatus();
          }
        }
      },
      () => {
        /* no catch needed */
      }
    );
  };

  render() {
    const {
      component,
      appState: { projectImportFeatureEnabled }
    } = this.props;
    const { lastAnalysisTask, lastExportTask, lastImportTask, status } = this.state;

    return (
      <div className="page page-limited" id="project-dump">
        <header className="page-header">
          <h1 className="page-title">{translate('project_dump.page')}</h1>
          <div className="page-description">
            {projectImportFeatureEnabled
              ? translate('project_dump.page.description')
              : translate('project_dump.page.description_without_import')}
          </div>
        </header>

        {status === undefined ? (
          <i className="spinner" />
        ) : (
          <div className="columns">
            <div className="column-half">
              <Export
                componentKey={component.key}
                loadStatus={this.poll}
                status={status}
                task={lastExportTask}
              />
            </div>
            <div className="column-half">
              <Import
                importEnabled={!!projectImportFeatureEnabled}
                analysis={lastAnalysisTask}
                componentKey={component.key}
                loadStatus={this.poll}
                status={status}
                task={lastImportTask}
              />
            </div>
          </div>
        )}
      </div>
    );
  }
}

export default withAppState(ProjectDumpApp);
