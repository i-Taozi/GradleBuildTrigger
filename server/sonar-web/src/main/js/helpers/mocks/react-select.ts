/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { GroupTypeBase, OptionProps } from 'react-select';

export function mockReactSelectOptionProps<
  OptionType,
  IsMulti extends boolean,
  GroupType extends GroupTypeBase<OptionType> = GroupTypeBase<OptionType>
>(
  data: OptionType,
  overrides?: OptionProps<OptionType, IsMulti, GroupType>
): OptionProps<OptionType, IsMulti, GroupType> {
  return {
    ...overrides,
    data
  } as OptionProps<OptionType, IsMulti, GroupType>;
}
