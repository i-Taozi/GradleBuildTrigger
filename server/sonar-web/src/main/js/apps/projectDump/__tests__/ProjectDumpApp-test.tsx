/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { getActivity } from '../../../api/ce';
import { getStatus } from '../../../api/project-dump';
import { mockComponent } from '../../../helpers/mocks/component';
import { mockAppState, mockDumpStatus, mockDumpTask } from '../../../helpers/testMocks';
import { waitAndUpdate } from '../../../helpers/testUtils';
import { TaskStatuses } from '../../../types/tasks';
import { ProjectDumpApp } from '../ProjectDumpApp';

jest.mock('../../../api/ce', () => ({
  getActivity: jest.fn().mockResolvedValue({ tasks: [] })
}));

jest.mock('../../../api/project-dump', () => ({
  getStatus: jest.fn().mockResolvedValue({})
}));

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', async () => {
  (getActivity as jest.Mock)
    .mockResolvedValueOnce({ tasks: [mockDumpTask()] })
    .mockResolvedValueOnce({ tasks: [mockDumpTask()] })
    .mockResolvedValueOnce({ tasks: [mockDumpTask()] });

  let wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot('loading');

  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot('loaded');

  wrapper = shallowRender({ appState: mockAppState({ projectImportFeatureEnabled: false }) });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot('loaded without import');
});

it('should poll for task status update', async () => {
  jest.useFakeTimers();

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  jest.clearAllMocks();

  const finalStatus = mockDumpStatus({ exportedDump: 'export-path' });
  (getStatus as jest.Mock)
    .mockResolvedValueOnce(mockDumpStatus())
    .mockResolvedValueOnce(finalStatus);
  (getActivity as jest.Mock)
    .mockResolvedValueOnce({ tasks: [mockDumpTask({ status: TaskStatuses.Pending })] })
    .mockResolvedValueOnce({ tasks: [mockDumpTask({ status: TaskStatuses.Success })] });

  wrapper.instance().poll();

  // wait for all promises
  await waitAndUpdate(wrapper);
  jest.runAllTimers();
  await waitAndUpdate(wrapper);

  expect(getStatus).toHaveBeenCalledTimes(2);
  expect(wrapper.state().status).toBe(finalStatus);
});

function shallowRender(overrides: Partial<ProjectDumpApp['props']> = {}) {
  return shallow<ProjectDumpApp>(
    <ProjectDumpApp
      appState={mockAppState({ projectImportFeatureEnabled: true })}
      component={mockComponent()}
      {...overrides}
    />
  );
}
