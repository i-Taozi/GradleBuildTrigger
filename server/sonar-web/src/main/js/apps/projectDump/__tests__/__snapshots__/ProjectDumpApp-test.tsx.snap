// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: loaded 1`] = `
<div
  className="page page-limited"
  id="project-dump"
>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      project_dump.page
    </h1>
    <div
      className="page-description"
    >
      project_dump.page.description
    </div>
  </header>
  <div
    className="columns"
  >
    <div
      className="column-half"
    >
      <Export
        componentKey="my-project"
        loadStatus={[Function]}
        status={Object {}}
        task={
          Object {
            "executedAt": "2020-03-12T12:22:20Z",
            "startedAt": "2020-03-12T12:20:20Z",
            "status": "SUCCESS",
            "submittedAt": "2020-03-12T12:15:20Z",
          }
        }
      />
    </div>
    <div
      className="column-half"
    >
      <Import
        analysis={
          Object {
            "executedAt": "2020-03-12T12:22:20Z",
            "startedAt": "2020-03-12T12:20:20Z",
            "status": "SUCCESS",
            "submittedAt": "2020-03-12T12:15:20Z",
          }
        }
        componentKey="my-project"
        importEnabled={true}
        loadStatus={[Function]}
        status={Object {}}
        task={
          Object {
            "executedAt": "2020-03-12T12:22:20Z",
            "startedAt": "2020-03-12T12:20:20Z",
            "status": "SUCCESS",
            "submittedAt": "2020-03-12T12:15:20Z",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render correctly: loaded without import 1`] = `
<div
  className="page page-limited"
  id="project-dump"
>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      project_dump.page
    </h1>
    <div
      className="page-description"
    >
      project_dump.page.description_without_import
    </div>
  </header>
  <div
    className="columns"
  >
    <div
      className="column-half"
    >
      <Export
        componentKey="my-project"
        loadStatus={[Function]}
        status={Object {}}
      />
    </div>
    <div
      className="column-half"
    >
      <Import
        componentKey="my-project"
        importEnabled={false}
        loadStatus={[Function]}
        status={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`should render correctly: loading 1`] = `
<div
  className="page page-limited"
  id="project-dump"
>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      project_dump.page
    </h1>
    <div
      className="page-description"
    >
      project_dump.page.description
    </div>
  </header>
  <i
    className="spinner"
  />
</div>
`;
